File: tools/lint_wiki.py
#!/usr/bin/env python3
from __future__ import annotations

import datetime as dt
import re
from collections import defaultdict
from pathlib import Path

from wiki_common import WIKI_ROOT, append_log, parse_date, slugify, split_frontmatter, today_iso


SKIP_ORPHAN_NAMES = {
    "CLAUDE.md",
    "index.md",
    "log.md",
    "me.md",
    "priorities.md",
    "mine.md",
    "theirs.md",
}


def wiki_files() -> list[Path]:
    files: list[Path] = []
    for path in WIKI_ROOT.rglob("*.md"):
        rel = path.relative_to(WIKI_ROOT)
        if rel.parts[0] == "raw":
            continue
        files.append(path)
    return sorted(files)


def wikilinks(text: str) -> list[str]:
    return re.findall(r"\[\[([^\]|#]+)", text)


def page_key(path: Path) -> str:
    return slugify(path.stem)


def action_due_dates(text: str) -> list[tuple[int, dt.date, str]]:
    found: list[tuple[int, dt.date, str]] = []
    patterns = [
        re.compile(r"- \[ \].*?\bdue:\s*(\d{4}-\d{2}-\d{2})", re.IGNORECASE),
        re.compile(r"- \[ \].*?@due\((\d{4}-\d{2}-\d{2})\)", re.IGNORECASE),
    ]
    for lineno, line in enumerate(text.splitlines(), start=1):
        for pattern in patterns:
            match = pattern.search(line)
            if not match:
                continue
            due = parse_date(match.group(1))
            if due:
                found.append((lineno, due, line.strip()))
    return found


def format_list(items: list[str]) -> str:
    if not items:
        return "- None found."
    return "\n".join(f"- {item}" for item in items)


def main() -> int:
    today = dt.date.fromisoformat(today_iso())
    files = wiki_files()
    bodies: dict[Path, str] = {}
    metadata: dict[Path, dict[str, str]] = {}
    missing_frontmatter: list[str] = []
    inbound: dict[str, set[Path]] = defaultdict(set)
    known_keys: dict[str, Path] = {}

    for path in files:
        text = path.read_text(encoding="utf-8")
        fm, body = split_frontmatter(text)
        bodies[path] = body
        metadata[path] = fm
        known_keys[page_key(path)] = path
        if path.name != "log.md" and not fm:
            missing_frontmatter.append(str(path.relative_to(WIKI_ROOT)))
        for link in wikilinks(text):
            inbound[slugify(Path(link).stem)].add(path)

    stale_threads: list[str] = []
    missing_decisions: list[str] = []
    orphan_pages: list[str] = []
    overdue_actions: list[str] = []
    stale_priorities: list[str] = []

    for path, fm in metadata.items():
        rel = path.relative_to(WIKI_ROOT)
        page_type = fm.get("type", "")
        if page_type == "thread" and fm.get("status", "open") == "open":
            opened = parse_date(fm.get("opened"))
            if opened and (today - opened).days > 30:
                stale_threads.append(f"{rel} opened {opened.isoformat()} ({(today - opened).days} days old)")

        if path.name == "priorities.md":
            last_updated = parse_date(fm.get("last_updated"))
            if not last_updated:
                stale_priorities.append("priorities.md has no `last_updated` date.")
            elif (today - last_updated).days > 30:
                stale_priorities.append(f"priorities.md last updated {last_updated.isoformat()} ({(today - last_updated).days} days old)")

        if rel.parts[0] in {"people", "projects", "teams", "meetings", "decisions", "threads"}:
            if path.name not in SKIP_ORPHAN_NAMES and page_key(path) not in inbound:
                orphan_pages.append(str(rel))

        if rel.parts[0] == "actions":
            for lineno, due, line in action_due_dates(path.read_text(encoding="utf-8")):
                if due < today:
                    overdue_actions.append(f"{rel}:{lineno} due {due.isoformat()} | {line}")

    decision_dir = WIKI_ROOT / "decisions"
    decision_keys = {page_key(path) for path in decision_dir.glob("*.md")}
    for path, body in bodies.items():
        for link in wikilinks(body):
            key = slugify(Path(link).stem)
            if re.match(r"\d{4}-\d{2}-\d{2}-", key) and key not in decision_keys and "decision" in link.lower():
                missing_decisions.append(f"{path.relative_to(WIKI_ROOT)} links to missing decision `[[{link}]]`")

    report = f"""# Wiki Lint — {today.isoformat()}

## Open Threads Older Than 30 Days

{format_list(stale_threads)}

## Stale Priorities

{format_list(stale_priorities)}

## Decisions Referenced But Missing

{format_list(sorted(set(missing_decisions)))}

## Orphan Pages

{format_list(orphan_pages)}

## Missing Frontmatter

{format_list(missing_frontmatter)}

## Overdue Action Items

{format_list(overdue_actions)}
"""

    output = WIKI_ROOT / "meta" / "lint" / f"lint-{today.isoformat()}.md"
    output.parent.mkdir(parents=True, exist_ok=True)
    output.write_text(report, encoding="utf-8")
    append_log("lint", today.isoformat())
    print(f"wrote {output.relative_to(WIKI_ROOT)}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/tools/lintWiki.ts
#!/usr/bin/env bun
import { readdir, readFile, writeFile } from "node:fs/promises";
import path from "node:path";
import {
  appendLog,
  dateDiffDays,
  parseIsoDate,
  slugify,
  splitFrontmatter,
  todayIso,
  wikiRoot,
} from "./common.js";

const skipDirs = new Set([".git", "dist", "docs", "meta", "node_modules", "raw", "tools"]);
const skipOrphanNames = new Set([
  "AGENTS.md",
  "CLAUDE.md",
  "index.md",
  "log.md",
  "me.md",
  "priorities.md",
  "mine.md",
  "theirs.md",
]);

type Args = {
  dryRun: boolean;
};

function parseArgs(argv: string[]): Args {
  const args: Args = { dryRun: false };
  for (const arg of argv) {
    if (arg === "--help" || arg === "-h") {
      console.log("usage: lintWiki [--dry-run]");
      process.exit(0);
    }
    if (arg === "--dry-run") {
      args.dryRun = true;
    } else {
      throw new Error(`Unknown argument: ${arg}`);
    }
  }
  return args;
}

async function markdownFiles(dir = wikiRoot): Promise<string[]> {
  const entries = await readdir(dir, { withFileTypes: true });
  const files: string[] = [];
  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      if (!skipDirs.has(entry.name)) {
        files.push(...(await markdownFiles(fullPath)));
      }
    } else if (entry.isFile() && entry.name.endsWith(".md")) {
      files.push(fullPath);
    }
  }
  return files.sort();
}

function wikilinks(text: string): string[] {
  return [...text.matchAll(/\[\[([^\]|#]+)/g)].map((match) => match[1] ?? "");
}

function pageKey(filePath: string): string {
  return slugify(path.basename(filePath, ".md"));
}

function actionDueDates(text: string): { lineNumber: number; due: Date; line: string }[] {
  const found: { lineNumber: number; due: Date; line: string }[] = [];
  const patterns = [
    /- \[ \].*?\bdue:\s*(\d{4}-\d{2}-\d{2})/i,
    /- \[ \].*?@due\((\d{4}-\d{2}-\d{2})\)/i,
  ];
  const lines = text.split(/\r?\n/);
  lines.forEach((line, index) => {
    for (const pattern of patterns) {
      const match = pattern.exec(line);
      const due = parseIsoDate(match?.[1]);
      if (due) {
        found.push({ lineNumber: index + 1, due, line: line.trim() });
      }
    }
  });
  return found;
}

function formatList(items: string[]): string {
  return items.length === 0 ? "- None found." : items.map((item) => `- ${item}`).join("\n");
}

async function main(): Promise<number> {
  const args = parseArgs(process.argv.slice(2));
  const today = parseIsoDate(todayIso());
  if (!today) {
    throw new Error("Could not parse current date");
  }

  const files = await markdownFiles();
  const bodies = new Map<string, string>();
  const metadata = new Map<string, Record<string, string>>();
  const inbound = new Map<string, Set<string>>();
  const missingFrontmatter: string[] = [];

  for (const filePath of files) {
    const text = await readFile(filePath, "utf8");
    const { metadata: fm, body } = splitFrontmatter(text);
    bodies.set(filePath, body);
    metadata.set(filePath, fm);
    if (!skipOrphanNames.has(path.basename(filePath)) && Object.keys(fm).length === 0) {
      missingFrontmatter.push(path.relative(wikiRoot, filePath));
    }
    for (const link of wikilinks(text)) {
      const key = slugify(path.basename(link, ".md"));
      const pages = inbound.get(key) ?? new Set<string>();
      pages.add(filePath);
      inbound.set(key, pages);
    }
  }

  const staleThreads: string[] = [];
  const missingDecisions: string[] = [];
  const orphanPages: string[] = [];
  const overdueActions: string[] = [];
  const stalePriorities: string[] = [];

  for (const [filePath, fm] of metadata.entries()) {
    const rel = path.relative(wikiRoot, filePath);
    const pageType = fm["type"] ?? "";
    if (pageType === "thread" && (fm["status"] ?? "open") === "open") {
      const opened = parseIsoDate(fm["opened"]);
      if (opened && dateDiffDays(today, opened) > 30) {
        staleThreads.push(
          `${rel} opened ${opened.toISOString().slice(0, 10)} (${dateDiffDays(today, opened)} days old)`,
        );
      }
    }

    if (path.basename(filePath) === "priorities.md") {
      const lastUpdated = parseIsoDate(fm["last_updated"]);
      if (!lastUpdated) {
        stalePriorities.push("priorities.md has no `last_updated` date.");
      } else if (dateDiffDays(today, lastUpdated) > 30) {
        stalePriorities.push(
          `priorities.md last updated ${lastUpdated.toISOString().slice(0, 10)} (${dateDiffDays(today, lastUpdated)} days old)`,
        );
      }
    }

    const topLevelDir = rel.split(path.sep)[0];
    if (
      topLevelDir &&
      ["people", "projects", "teams", "meetings", "decisions", "threads"].includes(topLevelDir)
    ) {
      if (!skipOrphanNames.has(path.basename(filePath)) && !inbound.has(pageKey(filePath))) {
        orphanPages.push(rel);
      }
    }

    if (topLevelDir === "actions") {
      const text = await readFile(filePath, "utf8");
      for (const { lineNumber, due, line } of actionDueDates(text)) {
        if (due < today) {
          overdueActions.push(
            `${rel}:${lineNumber} due ${due.toISOString().slice(0, 10)} | ${line}`,
          );
        }
      }
    }
  }

  const decisionKeys = new Set<string>();
  for (const filePath of files) {
    if (path.relative(wikiRoot, filePath).startsWith(`decisions${path.sep}`)) {
      decisionKeys.add(pageKey(filePath));
    }
  }

  for (const [filePath, body] of bodies.entries()) {
    for (const link of wikilinks(body)) {
      const key = slugify(path.basename(link, ".md"));
      if (/^\d{4}-\d{2}-\d{2}-/.test(key) && key.includes("decision") && !decisionKeys.has(key)) {
        missingDecisions.push(
          `${path.relative(wikiRoot, filePath)} links to missing decision \`[[${link}]]\``,
        );
      }
    }
  }

  const report = `# Wiki Lint — ${today.toISOString().slice(0, 10)}

## Open Threads Older Than 30 Days

${formatList(staleThreads)}

## Stale Priorities

${formatList(stalePriorities)}

## Decisions Referenced But Missing

${formatList([...new Set(missingDecisions)].sort())}

## Orphan Pages

${formatList(orphanPages)}

## Missing Frontmatter

${formatList(missingFrontmatter)}

## Overdue Action Items

${formatList(overdueActions)}
`;

  const output = path.join(wikiRoot, "meta", "lint", `lint-${today.toISOString().slice(0, 10)}.md`);
  if (args.dryRun) {
    console.log(report);
    return 0;
  }
  await writeFile(output, report, "utf8");
  await appendLog("lint", today.toISOString().slice(0, 10));
  console.log(`wrote ${path.relative(wikiRoot, output)}`);
  return 0;
}

main().then(
  (code) => {
    process.exitCode = code;
  },
  (error: unknown) => {
    console.error(error instanceof Error ? error.message : String(error));
    process.exitCode = 1;
  },
);
